### PR TITLE
Parallelize resolve.

### LINKED ISSUE #811: Slow builds in 2.0 - how does --python interact with --platform?
I want to build a pex that supports Linux armv7l and x86_64, on Python versions 3.5 through 3.7.

Before pex 2.0, I used

`--platform linux_armv7l --platform linux_x86_64 --python python3.5 --python python3.6 --python python3.7 --python-shebang "/usr/bin/env python3"`

And that produced a pex file meeting those requirements in about ten seconds.

As of 2.0 I'm not sure what I need to do. pex seems to resolve requirements for every combination of platform and python, so if I specify all six relevant `--platform` options (`--platform linux_armv7l-cp-35-m`, `--platform linux_x86_64-cp-36-m`, etc) plus all three `--python` options it takes **200s**.

For my use case I can simplify and specify just `--platform any-py-<35/36/37>-none` and the `--python`s. That takes **90s**.

Ultimately the dependencies are in this case resolved to py3-none-any wheels, so I figure I can probably get away with specifying `--platform any-py-35-none` and the `--python`s. That takes **60s**.

If I omit `--python` entirely with the single `--platform` arg, it takes about **15s** - close to the pre-2.0 time. For every `--python` I add, including the default interpreter, it adds another ~15s.

In all these cases the actual resolved and bundled dependencies are identical. What effect does `--python` actually have on the pex over and above the platform specifier? Is there anything I can do to get the build time down while ensuring that I'm producing a file that will work where it needs to?

### LINKED ISSUE #817: Cache sdist & local project builds
Currently, pex does not cache builds of sdists and local projects. Re-instate this cache - which was present in pre version 2 pex with bugs - in a safe way, invalidating the cache based on the content hash of sdists / local project directory trees.

This is part of the story for rectifying #811 in addition to caching installs (#815) and parallelizing resolves (#818).

### LINKED ISSUE #818: Parallelize resolve.
PEX has always performed a serial resolve, first resolving, then building, then installing. We can take advantage of the fact we shell out to pip for these three steps and use process parallelism for an easy way to improve cold cache pex build speed.

This is part of the story for rectifying #811 in addition to caching installs (#815) and builds (#817).

### PR DESCRIPTION
**Pull request by jsirois**
The three major phases of a pex resolve: download, build and install,
are delegated to pip subprocesses. As such, we can easily parallelize
these operations only needing to take care of shared portions of the
filesystem the processes might mutate. In the end this is only relevant
to the build and install phases which are natural points to cache
results in a shared filesystem cache. The install phase is already
cached in this way (#815) and we add caching for the build phase as
well, both utilizing (posix in the end) guarantees around `os.rename`.

The atomic shared directory updates are achieved with `AtomicDirectory`
and the subprocess parallelization is acheived using the new `jobs`
module and request / response data object pairs to coordinate parallel
pip jobs.

Example speedups:

+ PEX build using prebuilt wheels (the #811 case):
  ```
  pex \
    --platform=manylinux1-x86_64-cp-35-m \
    --platform=manylinux1-x86_64-cp-36-m \
    --platform=manylinux1-x86_64-cp-37-m \
    --platform=macosx-10.9-x86_64-cp-35-m \
    --platform=macosx-10.9-x86_64-cp-36-m \
    --platform=macosx-10.9-x86_64-cp-37-m \
    numpy==1.17.4 \
    --python-shebang "/usr/bin/env python3" \
    -o numpy-pex1.6.2.pex
  ```
  pex version | cold      | warm
  ------------|-----------|----------
  1.6.12      | 0m41.313s | 0m20.759s
  2.0.2       | 0m53.236s | 0m27.217s
  HEAD        | 0m32.336s | 0m17.596s

+ PEX build including sdists (wheel builds):
  ```
  pex \
    pantsbuild.pants==1.22.0 \
    -o pantsbuild.pants.pex
  ```
  pex version | cold      | warm
  ------------|-----------|----------
  1.6.12      | 0m49.101s | 0m11.788s
  2.0.2       | 1m15.602s | 1m2.476s
  HEAD        | 0m47.174s | 0m15.309s

Fixes #811
Fixes #817
Fixes #818

### CODE REVIEW

**Comment by jsirois** [comment]
I'd reccomend reviewing by reading the diffs in-order and not expanding the auto-collapsed `resolver.py` diff (it's large). This will allow you to familiarize yourself with `DistributionTarget` and `Jobs` which are two of the main players in the `resolver.py` diff. Once ready to look at the `resolver.py` diff, I'd focus on the `AtomicDirectory` class 1st, then skip to the `ResolveRequest` class and focus in its `resolve_distributions` method which is the old `resolve` function made parallel using jobs, AtomicDirectory and request / response object pairs for each resolve stage.

File: pex/bin/pex.py [hunk 1]
```diff
@@ -189,6 +190,15 @@ def configure_clp_pex_resolution(parser):
...
     help='Whether to transitively resolve requirements. Default: True')
 
+  group.add_option(
+    '-j', '--parallel',
+    metavar='JOBS',
+    dest='max_parallel_jobs',
+    type=int,
+    default=DEFAULT_MAX_JOBS,
+    help='The maximum number of parallel jobs to use when resolving, building and installing '
+         'distributions. [Default: %default]')
```
[the remark this answers is not in the document]
Comment (jsirois) [reply pex/bin/pex.py]: - "Increase the maximum number of parallel jobs to potentially improve the latency of the pex creation process at the expense of other processes on your system."

That would be a concise snippet that conveys why in the cases I can think of. The decrease why is the inverse so I leave out. Let me know if you find that helpful to add to the help string after consideringprior art in directly relevant tools:
```
$ make --help 2>&1 | grep jobs
  -j [N], --jobs[=N]          Allow N jobs at once; infinite jobs with no arg.
                              Don't start multiple jobs unless load is below N.
                              Synchronize output of parallel jobs by TYPE.
$ ninja --help 2>&1 | grep jobs
  -j N     run N jobs in parallel (0 means infinity) [default=10 on this system]
  -k N     keep going until N jobs fail (0 means infinity) [default=1]
  -l N     do not start new jobs if the load average is greater than N
$ clang --help 2>&1 | grep jobs
  -flto-jobs=<value>      Controls the backend parallelism of -flto=thin (default of 0 means the number of threads will be derived from the number of CPUs detected)
```

In other words, by implication, the why is widespread knowledge. That's not a great answer, since it still begs the question of how it became widespread since no tools I know of explain the why. The reason I suspect the why is not needed, is the tool is used by programmers and they generally understand:
1. My machine is melting down -> decrease parallelism.
2. This thing is slow -> increase.

File: pex/distribution_target.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+      return 'Target(platform={!r})'.format(self._platform)
+
+  def _tup(self):
+    return self._interpreter, self._platform
+
+  def __eq__(self, other):
+    return type(self) == type(other) and self._tup() == other._tup()
+
+  def __ne__(self, other):
+    return not self.__eq__(other)
```
[the remark this answers is not in the document]
Comment (jsirois) [reply pex/distribution_target.py]: Excellent, fixed.

File: pex/jobs.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+class Job(object):
+  """Represents a job spawned as a subprocess."""
+
+  class Error(Exception):
+    """Indicates that a Job exited non-zero."""
+
+  def __init__(self, command, process):
+    self._command = command
+    self._process = process
```
[the remark this answers is not in the document]
Comment (jsirois) [reply pex/jobs.py]: Docs added.

File: pex/jobs.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+    N.B.: This method is idempotent.
+
+    :raises: :class:`Job.Error` if the job exited non-zero.
+    """
+    self._process.wait()
+    self._check_returncode()
+
+  def communicate(self, input=None):
+    """communicates with the job sending any input data to stdin and collecting stdout and stderr.
```
[the remark this answers is not in the document]
Comment (jsirois) [reply pex/jobs.py]: I think repeating passthrough doc has its own host of issues. Now that the docs more clearly call out process is a subprocess.Popen, I've just added verbage to indicate communicate mirrors that API.

File: pex/bin/pex.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Eric-Arellano) [reply pex/bin/pex.py]: > "Increase the maximum number of parallel jobs to potentially improve the latency of the pex creation process at the expense of other processes on your system."

This is great! Very helpful. 

Rather than relying on the assumption that Pex users already know the why behind this option, this small snippet makes our tool more accessible for beginners, which is a big win in my book.

File: pex/jobs.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+      if e.errno != errno.ESRCH:
+        raise e
+
+  def _check_returncode(self):
+    if self._process.returncode != 0:
+      raise self.Error('Executing {} failed with {}'
+                       .format(' '.join(self._command), self._process.returncode))
+
+  def __str__(self):
+    return 'pid: {pid} -> {command}'.format(pid=self._process.pid, command=' '.join(self._command))
```
[the remark this answers is not in the document]
Comment (jsirois) [reply pex/jobs.py]: I'd tend to do that in `repr` but not str. If I want to print a user facing message and I bake internal type info in __str__ implementations, I now may communicate useless confusing information. Is your general stance that for user facing messages you should hand roll those at the sport where you emit them? Or does emitting internal type names not strike you as problematic? IE: If `Job` is a good name, it's a good name period?

File: pex/jobs.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+
+
+def spawn_python_job(args, env=None, interpreter=None, expose=None, **subprocess_kwargs):
+  """Spawns a python job.
```
[the remark this answers is not in the document]
Comment (jsirois) [reply pex/jobs.py]: This is another prior art thing, I've only ever heard spawn used in the async case for process interaction. The answer is also in the return type which is documented.

File: pex/jobs.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+
+    return cls(job=job, result_func=stdout_result_func)
+
+  def await_result(self):
+    """Waits for the spawned job to complete and returns its result."""
+    return self.result_func()
+
+  def kill(self):
+    """Terminates the spawned job if it's not already complete."""
+    self.job.kill()
```
[the remark this answers is not in the document]
Comment (jsirois) [reply pex/jobs.py]: Docs for job and result_func added.

File: pex/jobs.py [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+    spawned_job_queue.put(queue_done_sentinel)
+
+  spawner = Thread(name='PEX Parallel Job Spawner', target=spawn_jobs)
+  spawner.daemon = True
+  spawner.start()
+
+  error = None
+  while True:
+    item = spawned_job_queue.get()
+    if item is queue_done_sentinel:
```
Comment (Eric-Arellano) [inline pex/jobs.py]: One blank line between 189 and 190, along with a possible comment `# Handle end of the queue`, would help me to understand things. Lines 190 - 193 are a logical block related to "the queue is done", whereas line 189 is simply "get from the queue".

File: pex/jobs.py [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+  :raises: A `raise_type` exception if any individual job errors.
+  """
+  size = _sanitize_max_jobs(max_jobs)
+  TRACER.log('Spawning a maximum of {} parallel jobs to process:\n  {}'
+             .format(size, '\n  '.join(map(str, inputs))),
+             V=9)
+
+  stop = Event()
+  job_slots = BoundedSemaphore(value=size)
+  spawned_job_queue = Queue()
```
[the remark this answers is not in the document]
Comment (Eric-Arellano) [reply pex/jobs.py]: As discussed, this should actually be `Queue[Union[SpawnedJob, Exception, object]]`

File: pex/jobs.py [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+    item = spawned_job_queue.get()
+    if item is queue_done_sentinel:
+      if error:
+        raise error
+      return
+
+    try:
+      if isinstance(item, Exception):
+        error = item
+      elif error is not None:
```
Comment (Eric-Arellano) [inline pex/jobs.py]: Can't think of a better way to structure the code, so for now what would have helped me is a brief comment:

```python
      elif error is not None:  # i.e. this is not an exception, but there were prior exceptions
```

File: pex/jobs.py [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+        item.job.kill()
+      else:
```
Comment (Eric-Arellano) [inline pex/jobs.py]:
```python
      else:  # i.e. this is not an exception and there were no prior exceptions
```

File: pex/distribution_target.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jsirois) [reply pex/distribution_target.py]: Actually this is false:
> ... so `123 == DistributionTarget` would raise an exception.

It's users of the comparison method that can choose to raise if they encounter `NotImplemented` as the return value:

```python
$ python2.7
Python 2.7.17 (default, Oct 22 2019, 09:14:09) 
[GCC 9.2.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> class Example(object):
...   def __init__(self, data):
...     self._data = data
...   def __eq__(self, other):
...     if type(other) is not type(self):
...       return NotImplemented
...     return self._data == other._data
...   def __lt__(self, other):
...     if type(other) is not type(self):
...       return NotImplemented
...     return self._data < other._data
...   def __hash__(self):
...     return hash(self._data)
... 
>>> Example(42) == 42
False
>>> bool(NotImplemented)
True
>>> set([Example(42), 42])
set([<__main__.Example object at 0x7f2e4b125950>, 42])
>>> sorted([Example(42), 42])
[42, <__main__.Example object at 0x7f2e4b125910>]
>>> 
```

```python
$ python3.6
Python 3.6.9 (default, Jul  8 2019, 20:56:32) 
[GCC 9.1.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> class Example(object):
...   def __init__(self, data):
...     self._data = data
...   def __eq__(self, other):
...     if type(other) is not type(self):
...       return NotImplemented
...     return self._data == other._data
...   def __lt__(self, other):
...     if type(other) is not type(self):
...       return NotImplemented
...     return self._data < other._data
...   def __hash__(self):
...     return hash(self._data)
... 
>>> Example(42) == 42
False
>>> bool(NotImplemented)
True
>>> set([Example(42), 42])
{<__main__.Example object at 0x7fc5d7ef74e0>, 42}
>>> sorted([Example(42), 42])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '<' not supported between instances of 'int' and 'Example'
>>> 
```

File: pex/jobs.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jsirois) [reply pex/jobs.py]: I played with this and the idempotency behavior mirrored subprocess for python2.7 and 3 in all cases except for `kill` which is idempotent under 3 but not two, so I killed all the idempotency notes except for `kill`s and left the similar API except where noted comment in the class doc.

File: pex/jobs.py [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+                     raiseable value when any of the spawned jobs errors.
+  :returns: An iterator over the spawned job results as they come in.
+  :raises: A `raise_type` exception if any individual job errors.
+  """
+  size = _sanitize_max_jobs(max_jobs)
+  TRACER.log('Spawning a maximum of {} parallel jobs to process:\n  {}'
+             .format(size, '\n  '.join(map(str, inputs))),
+             V=9)
+
+  stop = Event()
```
[the remark this answers is not in the document]
Comment (jsirois) [reply pex/jobs.py]: The rename would be redundant since that that is the only use of `Event` objects, they are inter-thread signals. Added the comment instead.

File: pex/jobs.py [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+  job_slots = BoundedSemaphore(value=size)
+  spawned_job_queue = Queue()
+  queue_done_sentinel = object()
```
[the remark this answers is not in the document]
Comment (jsirois) [reply pex/jobs.py]: Fixed.

File: pex/jobs.py [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+
+  def spawn_jobs():
+    for input in inputs:
```
[the remark this answers is not in the document]
Comment (jsirois) [reply pex/jobs.py]: I did some renames instead.

File: pex/jobs.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (jsirois) [reply pex/jobs.py]: I went with `Queue[Union[SpawnedJob, Exception, Literal[done_sentinel]]]` even though `Literal` is experimental / will likely never support values like `object()` since `object` is clear as mud if you pause on it for a second.

File: pex/jobs.py (hunk 8 above)
Comment (jsirois) [reply pex/jobs.py]: I don't disagree, done.

File: pex/jobs.py (hunk 10 above)
Comment (jsirois) [reply pex/jobs.py]: Done.

File: pex/jobs.py (hunk 11 above)
Comment (jsirois) [reply pex/jobs.py]: This one I left out. The prior comment seems useful because of spatial closeness, the words and lines are just close enough together to cause confusion:
```python
...
  error = item
elif error is not None:
...
```

The rest of the logic stands on its own after the comment above, i.e.: exception, prior exception, nominal case.

File: pex/jobs.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Eric-Arellano) [reply pex/jobs.py]: Fair enough. Thanks!

File: pex/bin/pex.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jsirois) [reply pex/bin/pex.py]: I don't agree here, but done.

File: pex/jobs.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Eric-Arellano) [reply pex/jobs.py]: Thanks!

File: pex/jobs.py [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+  :type interpreter: :class:`PythonInterpreter`
+  :param expose: The names of any vendored distributions to expose to the spawned python process.
+  :type expose: list of str
+  :param subprocess_kwargs: Any additional :class:`subprocess.Popen` kwargs to pass through.
+  :returns: A job handle to the spawned python process.
+  :rtype: :class:`Job`
+  """
+  if expose:
+    subprocess_env = (env or os.environ).copy()
+    subprocess_env['__PEX_UNVENDORED__'] = '1'
```
[the remark this answers is not in the document]
Comment (jsirois) [reply pex/jobs.py]: Yeah - this is non-local magic. Commented with a link to the vendoring code bit that reads this.

File: pex/jobs.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (Eric-Arellano) [reply pex/jobs.py]: Sounds good. Thanks for adding a comment!

File: pex/jobs.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Eric-Arellano) [reply pex/jobs.py]: Ah much better. Thank you!

I'm not concerned about MyPy recognizing this as valid. We're not actually using MyPy. Rather, the reason I pitched a type comment is that it's now a standardized way in Python to express type information.

File: pex/jobs.py (hunk 11 above)
Comment (Eric-Arellano) [reply pex/jobs.py]: Sounds good. With the comment above, I agree with the decision for no comment here.

File: pex/distribution_target.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Eric-Arellano) [reply pex/distribution_target.py]: Hm, interesting. See https://medium.com/@s16h/pythons-notimplemented-type-2d720137bf41 for what happens. If `a.__eq__(b)` returns `NotImplemented`, according to this post Python will try `b.__eq__(a)`. If that too returns `NotImplemented`, then Python will use the default `__eq__` for any two objects.

So, `return NotImplemented` still does seem to be the right thing to do.

File: pex/resolver.py [hunk 16]
```diff
@@ -95,6 +126,401 @@ def parsed_platform(platform=None):
...
+  @property
+  def is_finalized(self):
+    return os.path.exists(self.target_dir)
+
+  def finalize(self):
+    if self.is_finalized:
+      return
+
+    try:
+      os.rename(self.work_dir, self.target_dir)
```
Comment (Eric-Arellano) [inline pex/resolver.py]: The PR description mentioned us exploiting a property of Posix for this to work - would be good to memorialize that here.

File: pex/resolver.py [hunk 17]
```diff
@@ -95,6 +126,401 @@ def parsed_platform(platform=None):
...
+        raise e
+
+
+class ResolveResult(namedtuple('ResolveResult', ['target', 'download_dir'])):
+  @staticmethod
+  def _is_wheel(path):
+    return os.path.isfile(path) and path.endswith('.whl')
+
+  def _iter_distribution_paths(self):
+    if os.path.exists(self.download_dir):
```
Comment (Eric-Arellano) [inline pex/resolver.py]: If `self.download_dir` does not exist, then this will never yield anything, meaning `build_requests` and `install_requests` will also never yield anything. 

Is there a use case where we expect `download_dir` to not exist? If that is not supposed to be supported, then we should raise an exception.

If we do expect it to exist, you can draw more attention to this edge case by inverting the conditional:

```python
def _iter_distribution_paths(self):
  if not os.path.exists(self.download_dir):
    return
  for distribution in ...
    yield ...
```

File: pex/resolver.py [hunk 18]
```diff
@@ -95,6 +126,401 @@ def parsed_platform(platform=None):
...
+  def _iter_requirements_requests(self, install_requests):
+    if self.is_installed:
+      # N.B.: Direct snip from the Environment docs:
+      #
+      #  You may explicitly set `platform` (and/or `python`) to ``None`` if you
+      #  wish to map *all* distributions, not just those compatible with the
+      #  running platform or Python version.
+      #
+      # Since our requested target may be foreign, we make sure find all distributions installed by
+      # explicitly setting both `python` and `platform` to `None`.
```
Comment (Eric-Arellano) [inline pex/resolver.py]: Thanks for pointing this out.

File: pex/jobs.py [hunk 19]
```diff
@@ -163,22 +194,23 @@ def execute_parallel(max_jobs, inputs, spawn_func, raise_type):
...
-          item = e
-        finally:
-          spawned_job_queue.put(item)
-    spawned_job_queue.put(queue_done_sentinel)
+    for item in inputs:
+      if stop.is_set():
+        break
+      job_slots.acquire()
+      try:
+        resut = spawn_func(item)
```
Comment (Eric-Arellano) [inline pex/jobs.py]: Typo: `s/resut/result`.

Overall, this block is much easier for me to understand now. Thank you for iterating on it! Looks good to me.

File: pex/jobs.py [hunk 20]
```diff
@@ -0,0 +1,207 @@
...
+  def spawn_jobs():
+    for input in inputs:
+      if not stop.is_set():
+        job_slots.acquire()
+        try:
+          item = spawn_func(input)
+        except Exception as e:
+          item = e
+        finally:
+          spawned_job_queue.put(item)
```
[the remark this answers is not in the document]
Comment (jsirois) [reply pex/jobs.py]: Fixed.

File: pex/resolver.py (hunk 16 above)
Comment (jsirois) [reply pex/resolver.py]: Done.

File: pex/pip.py [hunk 21]
```diff
@@ -137,28 +126,53 @@ def download_distributions(target,
...
+    '--no-deps',
+    '--no-index',
+    '--only-binary', ':all:',
+    '--target', install_dir
+  ]
+
+  interpreter = target.get_interpreter()
+  if target.is_foreign:
+    if compile:
+      raise ValueError('Cannot compile bytecode for {} using {}.'.format(wheel, interpreter))
```
[the remark this answers is not in the document]
Comment (jsirois) [reply pex/pip.py]: Done.

File: pex/resolver.py (hunk 17 above)
Comment (jsirois) [reply pex/resolver.py]: Download dir will not exist for one common case, building a pex from a local project with no `install_requires` and one strange but historically important case, running `pex` (running the pex tool with no args at all which builds a full temporary pex with no requirements and drops you into an interpreter isolted from all site-packages). 

Early return it is - but our tastes here do diverge.

File: pex/resolver.py [hunk 22]
```diff
@@ -95,6 +126,401 @@ def parsed_platform(platform=None):
...
+  @classmethod
+  def for_target_dir(cls, target_dir):
+    return cls(work_dir='{}.{}'.format(target_dir, uuid4().hex), target_dir=target_dir)
+
+  @property
+  def is_finalized(self):
+    return os.path.exists(self.target_dir)
+
+  def finalize(self):
+    if self.is_finalized:
```
Comment (cosmicexplorer) [inline pex/resolver.py]: For the `finalize()` method here, it seems valid to perform this check without explicit inter-process locking, but I'm concerned that propagating the result of `.is_finalized` elsewhere may make things harder to reason about (unless we're absolutely sure that the resulting directory will never be moved or deleted). In particular, for this segment later on:
```python
      if not build_result.is_built:
        TRACER.log('Building {} to {}'.format(build_request.source_path, build_result.dist_dir))
        unsatisfied_build_requests.append(build_request)
      else:
        TRACER.log('Using cached build of {} at {}'
                   .format(build_request.source_path, build_result.dist_dir))
        install_requests.extend(build_result.finalize_build())
```
I think this code seems mostly legitimate, but what's a little confusing to me is how `AtomicDirectory#finalize()` will no-op if `.is_finalized` is true, but then in this excerpt, `BuildResult` will immediately call `.finalize_build()`, which calls `self.atomic_dir.finalize()` (which would then be a no-op).

One possible way to refactor this might be to convert the `.finalized` check into an `@contextmanager`, or perhaps to provide some hook method when constructing an `AtomicDirectory` which is run when the finalization occurs.

I'm not super opinionated about this, and I think memorializing the kind of guarantee we want in the `finalized()` method and/or the `.is_finalized` property like @Eric-Arellano suggested would be sufficient here.

File: pex/resolver.py [hunk 23]
```diff
@@ -95,6 +126,401 @@ def parsed_platform(platform=None):
...
+    spawn_install = functools.partial(self._spawn_install, installed_wheels_dir)
+    to_install = []
+
+    to_calculate_requirements_for = []
+
+    # 1. Resolve sdists and wheels.
+    with TRACER.timed('Resolving for:\n  '.format('\n  '.join(map(str, to_resolve)))):
+      for resolve_result in self._run_parallel(inputs=to_resolve,
+                                               spawn_func=spawn_resolve,
+                                               raise_type=Unsatisfiable):
```
[the remark this answers is not in the document]
Comment (jsirois) [reply pex/resolver.py]: Agreed on all points. Note though the shape of the steps is roughly:
+ downloads: 1-4 
+ builds: 1-10
+ installs: 1-100

For downloads, you may have 1-2 interpreters (python 2.7 + python 3) and -2 platforms.
For installs projects pulling in a transitive dep graph of 100 wheels is not unreasonable, but things probably don't get alot bigger than this either.
For builds, this will be a subset of the installs and it seems like (handwave), most pypi projects have prebuilt wheels for standard platforms and a 10% of my dependencies require local builds is about right for the repos I have access to.

File: pex/resolver.py [hunk 24]
```diff
@@ -95,6 +126,401 @@ def parsed_platform(platform=None):
...
+
+    # Dedup by wheel name; e.g.: only install universal wheels once even though they'll get
+    # downloaded / built for each interpreter or platform.
+    install_requests_by_wheel_file = OrderedDict()
+    for install_request in to_install:
+      install_requests = install_requests_by_wheel_file.setdefault(install_request.wheel_file, [])
+      install_requests.append(install_request)
+
+    representative_install_requests = [
+      requests[0] for requests in install_requests_by_wheel_file.values()
```
[the remark this answers is not in the document]
Comment (jsirois) [reply pex/resolver.py]: It's the `representative_install_request` for the given wheel file.

Concretely, if the resolve is over 2 targets, say two local interpreters - `python2.7` and `python3.6`, and the resolve hits a dependency that is a universal wheel like `ansicolors-1.1.8-py2.py3-none-any.whl`, we'll have:
```
OrderedDict([('ansicolors-1.1.8-py2.py3-none-any.whl',
              [InstallRequest(target=Target(interpreter='/usr/bin/python2.7'), wheel_path='/tmp/abc123/resolved_dists/ansicolors-1.1.8-py2.py3-none-any.whl', fingerprint='828462c81afbaa71068f55aaa6a5ea1d1647656f'),
               InstallRequest(target=Target(interpreter='/usr/bin/python3.6'), wheel_path='/tmp/abc123/resolved_dists/ansicolors-1.1.8-py2.py3-none-any.whl', fingerprint='828462c81afbaa71068f55aaa6a5ea1d1647656f')]),
             ('msgpack_python-0.5.6-cp27-cp27m-linux_x86_64.whl',
              [InstallRequest(target=Target(interpreter='/usr/bin/python2.7'), wheel_path='/tmp/abc123/resolved_dists/msgpack_python-0.5.6-cp27-cp27m-linux_x86_64.whl', fingerprint='7bbb4ce93159d76b7054e5372390cd85ccbc95e3')]),
             ('msgpack_python-0.5.6-cp36-cp36m-linux_x86_64.whl',
              [InstallRequest(target=Target(interpreter='/usr/bin/python3.6'), wheel_path='/tmp/abc123/resolved_dists/msgpack_python-0.5.6-cp36-cp36m-linux_x86_64.whl', fingerprint='7b7ffb31fb338973aaa3285b9630f8d5ea63ffda')])])
```

To install `'ansicolors-1.1.8-py2.py3-none-any.whl` we need only pick a single ([0]) `InstallRequest` from the list. This `InstallRequest` is "representative".

File: pex/resolver.py (hunk 22 above)
Comment (jsirois) [reply pex/resolver.py]: > One possible way to refactor this might be to convert the `.finalized` check into an `@contextmanager`

That's done in https://github.com/jsirois/pex/pull/1 which I'll be making into a PR here as soon as I land this PR - which is based upon.

File: pex/jobs.py [hunk 25]
```diff
@@ -0,0 +1,243 @@
...
+_ABSOLUTE_MAX_JOBS = _CPU_COUNT * 2
+
+
+DEFAULT_MAX_JOBS = _CPU_COUNT
+"""The default maximum number of parallel jobs PEX should use."""
+
+
+def _sanitize_max_jobs(max_jobs=None):
+  assert max_jobs is None or isinstance(max_jobs, int)
+  if max_jobs is None or max_jobs <= 0:
```
[the remark this answers is not in the document]
Comment (jsirois) [reply pex/jobs.py]: Perhaps, but as of now pex has no warning system, only -v(vvvvvvvvv). Just below the actual number used is logged at maximally fine level (9).

File: pex/resolver.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (cosmicexplorer) [reply pex/resolver.py]: Ok, that's great! Thanks! Reviewing the few lines before, I feel that the existing comment explains this sufficiently!

File: pex/jobs.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (cosmicexplorer) [reply pex/jobs.py]: Ok, great! I generally interpret an argument named like `--max-parallel-jobs` as a suggestion to the tool instead of a hard line, so the sanitizing here sounds quite reasonable.

File: pex/resolver.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (cosmicexplorer) [reply pex/resolver.py]: Thank you for this context!!

File: pex/resolver.py (hunk 22 above)
Comment (jsirois) [reply pex/resolver.py]: Now up for review in #821 and the @contextmanager is here: https://github.com/pantsbuild/pex/pull/821/files#diff-4895aa16756686554f1906f607a1296eR272
